### PR TITLE
`JavaNetClient` does not need `Async`

### PR DESCRIPTION
**Pull request by armanbilge**
`Sync` is sufficient, but making the change bincompatibly on `series/0.23` is not possible. Update: nevermind it was 😇 

Not sure how useful this is, but for example in our docs we _could_ use it to build a `Client[SyncIO]` that we can `unsafeRunSync()` without an implicit `IORuntime`. But that might be confusing.

### CODE REVIEW

**Comment by armanbilge** [comment]
Oh hang on, this is interesting.
```
[error] /home/runner/work/http4s/http4s/client/jvm/src/main/scala/org/http4s/client/JavaNetClientBuilder.scala:174:9: error: [Http4sFs2Linters.noFs2SyncCompiler] fs2's Sync compiler should be avoided. Usually this means a Sync constraint needs to be changed to Concurrent or upgraded to Async.
[error]         req.body
[error]         ^
[error] /home/runner/work/http4s/http4s/client/jvm/src/main/scala/org/http4s/client/JavaNetClientBuilder.scala:183:13: error: [Http4sFs2Linters.noFs2SyncCompiler] fs2's Sync compiler should be avoided. Usually this means a Sync constraint needs to be changed to Concurrent or upgraded to Async.
[error]             req.body
[error]             ^
```

**Comment by Quafadas** [comment]
Huh, interesting. If I read this right, then the actual request machinery is fine with Sync... 

... but the request _body parsing_ is done using FS2 machinery. The http4s project itself is setup to enforce good discipline in potentially concurrent operations, and so it complains about the Sync constraint?

**Comment by armanbilge** [comment]
Yes, well put :)

This is a tricky one. When the `Sync` compiler is used, instead of the `Concurrent` (`Async`) compiler, then the `Stream` loses the ability to be interrupted.

https://github.com/typelevel/fs2/blob/f6085648752a191ad72b0ad759f85d8ed6412631/core/shared/src/test/scala/fs2/StreamInterruptSuite.scala#L368-L372

We are not relying on interruption.
https://github.com/http4s/http4s/blob/20ce97918d4b1d023b2f0940ed0b2f0d65a13abe/client/jvm/src/main/scala/org/http4s/client/JavaNetClientBuilder.scala#L174-L177

However, the user-provided stream could still be relying on interruption.

Additionally, `SyncIO` itself is uncancelable.
https://github.com/typelevel/cats-effect/blob/a6a101c99a2878191900b84aec98d87d5ffac60f/core/shared/src/main/scala/cats/effect/SyncIO.scala#L579-L582

Lastly, despite the compile-time warning, FS2 will upgrade to the `Concurrent` compiler at runtime when possible.
- https://github.com/typelevel/fs2/pull/2898

So this is unlikely to be a problem in practice, but I think it could still be a footgun in some edge cases. Maybe we need a unit test to explore that situation.

**Comment by armanbilge** [comment]
> However, the user-provided stream could still be relying on interruption.

Actually, I think the only way this can be true, is if the effect for the user-provided stream has a `Concurrent` instance available. If that is the case, then ultimately the client will have to be run with an `Async`, and FS2 will upgrade the compiler at runtime.

There is still a footgun here in very very edge-casey situations (i.e. you have a stream relying on concurrent interruption and you wrap the `Async` instance in a `Sync` instance before passing it the `JavaNetClientBuilder`).

So all in all, I think it's safe to suppress the warning.

**Comment by rossabaker** [comment]
Yes, I'd opt out of the warning here.

The only reason I would consider using this client is for explorations in a REPL, where spawning threads has a habit of locking things up.  Its use is quite narrow, but `SyncIO` is a natural fit for that use.

File: client/jvm/src/main/scala/org/http4s/client/JavaNetClientBuilder.scala [hunk 1]
```diff
@@ -56,7 +55,7 @@ sealed abstract class JavaNetClientBuilder[F[_]] private (
     val proxy: Option[Proxy],
     val hostnameVerifier: Option[HostnameVerifier],
     val sslSocketFactory: Option[SSLSocketFactory],
-)(implicit protected val F: Async[F])
+)(implicit protected val F: Sync[F])
```
Comment (armanbilge) [inline client/jvm/src/main/scala/org/http4s/client/JavaNetClientBuilder.scala:58]: Actually I missed that this is a `protected val` in a `sealed` class. So we can make this change on 0.23 actually, we'll just have to filter it.

**Review by isomarcte — APPROVED** [review]

**Comment by isomarcte** [comment]
I love that last commit message ;)

**Review by rossabaker — APPROVED** [review]
